### ISSUE TITLE
Forest 1.3.1 crash on N

### ISSUE THREAD

**Issue by mtrakal** [couchbase/couchbase-lite-android#984]
We just release new app version, don't change CBL/Forest version (previous and new version use same version `1.3.1-16`).
But after upgrade a lot of people get this error.

It's caused by `com.couchbase.lite.DatabaseUpgrade$1Pair.revID`, so maybe users which upgrade from Android 6 to Android 7? (On Android 6 we use SQLite, on Android 7 use ForestDB)
**All crashes comes from Android 7.**

```
    compile 'com.couchbase.lite:couchbase-lite-android:1.3.1-16'
    compile 'com.couchbase.lite:couchbase-lite-android-forestdb:1.3.1-16'
```

```
Fatal Exception: java.lang.RuntimeException: Unable to create application com.droid4you.application.wallet.Application: java.lang.RuntimeException: java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.couchbase.lite.DatabaseUpgrade$1Pair.revID' on a null object reference
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5364)
       at android.app.ActivityThread.-wrap2(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6077)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by java.lang.RuntimeException: java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.couchbase.lite.DatabaseUpgrade$1Pair.revID' on a null object reference
       at com.couchbase.lite.store.ForestDBStore.runInTransaction(ForestDBStore.java:301)
       at com.couchbase.lite.Database.runInTransaction(Database.java:570)
       at com.couchbase.lite.DatabaseUpgrade.importData(DatabaseUpgrade.java:111)
       at com.couchbase.lite.Database.open(Database.java:1316)
       at com.couchbase.lite.Manager.openDatabase(Manager.java:335)
       at com.yablohn.internal.CouchDbProvider.connectDatabase(CouchDbProvider.java:294)
       at com.yablohn.internal.CouchDbProvider.init(CouchDbProvider.java:267)
       at com.yablohn.internal.CouchDbProvider.(CouchDbProvider.java)
       at com.yablohn.internal.Yablohn.initialize(Yablohn.java:17)
       at com.droid4you.application.wallet.Application.onCreate(Application.java:188)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5361)
       at android.app.ActivityThread.-wrap2(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6077)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by java.lang.NullPointerException: Attempt to read from field 'java.lang.String com.couchbase.lite.DatabaseUpgrade$1Pair.revID' on a null object reference
       at com.couchbase.lite.DatabaseUpgrade.importDoc(DatabaseUpgrade.java:255)
       at com.couchbase.lite.DatabaseUpgrade.access$000(DatabaseUpgrade.java:38)
       at com.couchbase.lite.DatabaseUpgrade$1.run(DatabaseUpgrade.java:118)
       at com.couchbase.lite.store.ForestDBStore.runInTransaction(ForestDBStore.java:297)
       at com.couchbase.lite.Database.runInTransaction(Database.java:570)
       at com.couchbase.lite.DatabaseUpgrade.importData(DatabaseUpgrade.java:111)
       at com.couchbase.lite.Database.open(Database.java:1316)
       at com.couchbase.lite.Manager.openDatabase(Manager.java:335)
       at com.yablohn.internal.CouchDbProvider.connectDatabase(CouchDbProvider.java:294)
       at com.yablohn.internal.CouchDbProvider.init(CouchDbProvider.java:267)
       at com.yablohn.internal.CouchDbProvider.(CouchDbProvider.java)
       at com.yablohn.internal.Yablohn.initialize(Yablohn.java:17)
       at com.droid4you.application.wallet.Application.onCreate(Application.java:188)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5361)
       at android.app.ActivityThread.-wrap2(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6077)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

**Comment by hideki** [comment]
`NullPointerException` is thrown from following line of codes because of `ancestor.revID` is `null`. 
https://github.com/couchbase/couchbase-lite-java-core/blob/fc5723da1e9dcefd2e698d2a511887d132131d33/src/main/java/com/couchbase/lite/DatabaseUpgrade.java#L255

As @mtrakal wrote, NPE occurs during migration from SQLite to ForestDB.
https://github.com/couchbase/couchbase-lite-java-core/blob/996396070b3a666f6c554c310c610f206f1d33d4/src/main/java/com/couchbase/lite/Database.java#L1223-L1236

https://github.com/couchbase/couchbase-lite-java-core/blob/996396070b3a666f6c554c310c610f206f1d33d4/src/main/java/com/couchbase/lite/Database.java#L1312-L1324

**Comment by hideki** [comment]
@mtrakal,
Error occurs during database upgrade from SQLite to ForestDB (1.3.1-16). From the log, I don't know what version of CBL users who encountered the issue use. Can you collect this information?

**Comment by hideki** [comment]
Note:
Most of tests for `DatabaseUpgrader` is covered by testing `Manager.replaceDatabase(...)` method, not by testing `Database.open(...)` method.

**Comment by mtrakal** [comment]
Hi @hideki ,

previous user version must be `1.2.1` we use this CBL version from final release (april?). So, every user which update to latest our app version with `1.3.1-16` had `1.2.1`

**Comment by hideki** [comment]
Hi @mtrakal,

I tried to reproduce this issue with using [ToDo Lite sample you provide](https://github.com/couchbase/couchbase-lite-android/issues/974#issuecomment-242383878).

I tested two scenarios.

Scenario 1: CBL 1.2.1 or earlier does not work with Android N, So used older version of Android.
- Setup ToDo app with SQLite storage type CBL v1.2.1
- Launch ToDo app on Android API 19
- Create some lists and tasks
- Modify ToDo app with ForestDB 1.3.1-47 (latest)
- Relaunch ToDo app.
Results:
- I can not observe any Exceptions with above test

Scenario 2: Use Android N
- Setup ToDo app with SQLite storage type CBL v1.3.1-47 (latest)
- Launch ToDo app on Android N
- Create some lists and tasks
- Modify ToDo app with ForestDB 1.3.1-47 (latest)
- Relaunch ToDo app.
Results:
- I can not observe any Exceptions with above test

Both scenarios, I can not reproduce the issue. Can you still reproduce this issue?

**Comment by mtrakal** [comment]
Hi @hideki ,

this issue I never reproduce. I saw it every time only from users. 

Bug on https://github.com/couchbase/couchbase-lite-android/issues/974 I can still reproduce on 1.3.1-47 and 1.3.2-4 too on Android 6 and Android 7.

We plan to release new version (this week) for our beta testers, we will use latest CBL. So I'll inform you here after.

**Comment by hideki** [comment]
@mtrakal,
Thank you for your updates.
Fix for #974 is still in review. https://github.com/couchbaselabs/cbforest/pull/109
Thanks!

**Comment by djpongh** [comment]
Unfortunately we can't reproduce so nothing we can do here. Please reopen if it happens again.
